Flink: Update to KUDO version 0.3.0 (#20)
kudobuilder/frameworks: repository/flink/docs/demo/financial-fraud/README.md

# Flink Financial Fraud Demo

> This demo follows the outline provided by [DCOS's](https://github.com/dcos/demos/tree/master/flink-k8s/1.11) demo

Reviewer: There is an updated version of this demo whose URL is https://github.com/dcos/demos/tree/master/flink-k8s/1.13

Author: Looking over https://github.com/dcos/demos/pull/86 it seems there are overall minor wording changes. Was there anything in particular note worthy that we should add in this demo here as well?

Reviewer: I think @gkleiman caught most of - and made similar recommendations to - anything wording related.

### Architecture

![Financial transaction processing demo architecture](https://github.com/dcos/demos/raw/master/flink-k8s/1.11/img/kafka-flink-arch.png)

This demo implements a data processing infrastructure with KUDO that is able to spot money laundering. In the context of money laundering, we  want to detect amounts larger than $10.000 transferred between two accounts, even if that amount is split into many small batches.  See also [US](https://www.fincen.gov/history-anti-money-laundering-laws) and [EU](http://eur-lex.europa.eu/legal-content/EN/TXT/?uri=CELEX%3A32015L0849) legislation and regulations on this topic for more information.

The architecture follows more or less the [SMACK stack architecture](https://mesosphere.com/blog/smack-stack-new-lamp-stack/):

> - Events: Event are being generated by a small [GOLANG generator](https://github.com/dcos/demos/blob/master/flink/1.11/generator/generator.go). The events are in the form 'Sunday, 23-Jul-17 01:06:47 UTC;66;26;7810', where the first field '23-Jul-17 01:06:47 UTC' represents the (increasing) timestamp of transactions; the second field '66' represent the sender account; the third field the receiver account; and the fourth field represent the dollar amount transferred during that transaction.

Reviewer 2: ```suggestion
- Events: Event are being generated by a small [Golang generator](https://github.com/dcos/demos/blob/master/flink/1.11/generator/generator.go). The events are in the form 'Sunday, 23-Jul-17 01:06:47 UTC;66;26;7810', where the first field '23-Jul-17 01:06:47 UTC' represents the (increasing) timestamp of transactions; the second field '66' represent the sender account; the third field the receiver account; and the fourth field represent the dollar amount transferred during that transaction.
```

- Ingestion: The generated events are being ingested and buffered by a Kafka queue with the default topic 'transactions'. Being a Microservice we will deploy the data-generator on kubernetes.
- Stream Processing: As we require fast response times, we use Apache Flink as a Stream processor running the [FinancialTransactionJob](https://github.com/dcos/demos/tree/master/flink/1.10/flink-job/src/main/java/io/dcos).
- Storage: Here we diverge a bit from the typical SMACK stack setup and don't write the results into a Datastore such as Apache Cassandra. Instead we write the results again into a Kafka Stream (default: 'fraud'). Note, that Kafka also offers data persistence for all unprocessed events.
- Actor: In order to view the results we use again a small [Golang viewer](https://github.com/dcos/demos/blob/master/flink/1.11/actor/actor_viewer.go) which simply reads and displays the results from the output Kafka stream. Being a Microservice we will deploy the viewer on kubernetes.

## Prerequisites

Before you get started:

> - Have KUDO run in your cluster [ [Getting Started](https://kudo.dev/docs/getting-started/) ]

Reviewer 2: ```suggestion
- Have KUDO installed on your cluster [ [Getting Started](https://kudo.dev/docs/getting-started/) ]
```

- Have the `zookeeper` Framework with `0.1.0` as FrameworkVersion installed 
    - Use the KUDO CLI with the following command:
        ```bash
        $ kubectl kudo install zookeeper --package-version=0.1.0
        framework.kudo.k8s.io/v1alpha1/zookeeper created
        frameworkversion.kudo.k8s.io/v1alpha1/zookeeper-0.1.0 created
        No Instance tied to this "zookeeper" version has been found. Do you want to create one? (Yes/no) no

        ```

>         *Note: Don't forget to say "**no**" when prompted, we just want to install the Framework and FrameworkVersion objects but not instantiated an Instance object*

Reviewer 2: ```suggestion
        *Note: Don't forget to say "**no**" when prompted, we just want to install the Framework and FrameworkVersion objects but we don't want to instantiate an Instance object*
```

- Have the `kafka` Framework with `0.1.0` as FrameworkVersion installed 
    - Use the KUDO CLI with the following command:
        ```bash
        $ kubectl kudo install kafka --package-version=0.1.0
        framework.kudo.k8s.io/v1alpha1/kafka created
        frameworkversion.kudo.k8s.io/v1alpha1/kafka-0.1.0 created
        No Instance tied to this "kafka" version has been found. Do you want to create one? (Yes/no) no

        ```

>         *Note: Don't forget to say "**no**" when prompted, we just want to install the Framework and FrameworkVersion objects but not instantiated an Instance object*

Reviewer 2: ```suggestion
        *Note: Don't forget to say "**no**" when prompted, we just want to install the Framework and FrameworkVersion objects but we don't want to instantiate an Instance object*
```

DOWNLOAD_URL: https://downloads.mesosphere.com/dcos-demo/flink/flink-job-1.0.jar FILE: flink-job-1.0.jar JOBMANAGER: demo-flink-jobmanager
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
(1/7) Installing ca-certificates (20171114-r3)
(2/7) Installing nghttp2-libs (1.32.0-r0)
(3/7) Installing libssh2 (1.8.0-r3)
(4/7) Installing libcurl (7.61.1-r1)
(5/7) Installing curl (7.61.1-r1)
(6/7) Installing oniguruma (6.8.2-r0)
(7/7) Installing jq (1.6_rc1-r1)
Executing busybox-1.28.4-r2.trigger
Executing ca-certificates-20171114-r3.trigger
OK: 15 MiB in 24 packages
{"filename":"/tmp/flink-web-1324b551-a734-43fe-824b-396d7760647c/flink-web-upload/684b6919-9e66-4064-94d9-22641c9c2fb1_flink-job-1.0.jar","status":"success"}Thu Jan 24 04:59:15 UTC 2019
No uploaded jar detected
=====================
Thu Jan 24 04:59:20 UTC 2019
Found jar 684b6919-9e66-4064-94d9-22641c9c2fb1_flink-job-1.0.jar
RESPONSE: null
SUBMITTED JOB!
```

To get the fraud output from the actor:

```bash
kubectl logs $(kubectl get pod -l step=act -o jsonpath="{.items[0].metadata.name}")
```

> Congratulations, you just installed an high available Flink cluster that runs a financial fraud detection job!

Reviewer 2: ```suggestion
Congratulations, you just installed a highly available Flink cluster that runs a financial fraud detection job!
```
